kernel/atomic_c.c

```diff
@@ -30,6 +30,30 @@
  */
 static struct k_spinlock lock;
 
+/* For those rare CPUs which support user mode, but not native atomic
+ * operations, the best we can do for them is implement the atomic
+ * functions as system calls, since in user mode locking a spinlock is
+ * forbidden.
+ */
+#ifdef CONFIG_USERSPACE
+#include <syscall_handler.h>
+
+#define ATOMIC_SYSCALL_HANDLER_TARGET(name) \
+	Z_SYSCALL_HANDLER(name, target) { \
+		Z_OOPS(Z_SYSCALL_MEMORY_WRITE(target, sizeof(atomic_t))); \
+		return z_impl_##name((atomic_t *)target); \
+	}
+
+#define ATOMIC_SYSCALL_HANDLER_TARGET_VALUE(name) \
+	Z_SYSCALL_HANDLER(name, target, value) { \
+		Z_OOPS(Z_SYSCALL_MEMORY_WRITE(target, sizeof(atomic_t))); \
+		return z_impl_##name((atomic_t *)target, value); \
+	}
+#else
+#define ATOMIC_SYSCALL_HANDLER_TARGET(name)
```

Comment:
Without CONFIG_USERSPACE no 'common' ATOMIC_OPERATION_C implementation required?
Maybe a #define ATOMIC_SYSCALL_HANDLER_TARGET(name)    z_impl_##name ?

Reply:
> Without CONFIG_USERSPACE no 'common' ATOMIC_OPERATION_C implementation required?

Yes it is, you still need logic for atomic operations from supervisor mode, even if no userspace is enabled. For arches not using custom code or the builtins: all RiscV32, all Nios II, some xtensa, Cortex-M0, etc

> Maybe a #define ATOMIC_SYSCALL_HANDLER_TARGET(name) z_impl_##name ?

That isn't necessary, it can just be blank. Build time magic enforces that supervisor mode calls to these APIs jump directly to the implementations, skipping software exception and handler functions. It's transparent with no overhead if CONFIG_USERSPACE=n. 

In a build directory after a kernel compile, see:

    zephyr/include/generated/syscall_macros.h
    zephyr/include/generated/syscalls/*.h

Reply:
Also see: https://docs.zephyrproject.org/latest/reference/kernel/usermode/syscalls.html